alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AbstractOrienting.kt

```diff
@@ -0,0 +1,417 @@
+     */
+    private lateinit var subdestination: P
+    /*
+     * The edge the pedestrian is moving towards.
+     */
+    private lateinit var targetDoor: F
+
+    private enum class State {
+        START,
+        NEW_ROOM,
+        /*
+         * There is one unusual case in which the pedestrian cannot locate itself
+         * inside any room of the environment's graph, and decides to try to reach
+         * the closest crossing possible. In such case, the state will be MOVING_TO
+         * _DOOR, but the [currRoom] variable won't be initialised yet.
+         */
+        MOVING_TO_DOOR,
+        /*
+         * Adjacent rooms are not guaranteed to be exactly geometrically adjacent,
+         * there can be some distance between them. Consequently, the pedestrian could
+         * find itself in a situation in which it is crossing a door but cannot locate
+         * itself inside any room. Such case is identified by CROSSING_DOOR: in this case,
+         * the pedestrian just proceeds towards the center of the room it wants to reach.
+         */
+        CROSSING_DOOR,
+        MOVING_TO_FINAL,
+        ARRIVED
+    }
+    private var state: State = State.START
+
+    private fun onStart() {
+        when {
+            environment.graph().nodeContaining(currentPosition) != null -> state = State.NEW_ROOM
+            /*
+             * If the pedestrian cannot locate itself inside any room (unusual
+             * condition), it tries to reach the closest door/passage in order
+             * to enter one. If this isn't possible, it simply won't move.
+             */
+            else -> moveToClosestDoor(ifNoDoorCanBeFound = { state = State.ARRIVED })
+        }
+    }
+
+    private fun inNewRoom() {
+        val newRoom = environment.graph().nodeContaining(currentPosition)
+            ?: throw IllegalStateException("pedestrian is not inside a room")
+        pedestrian.registerVisit(newRoom)
+        when {
+            ::nextRoom.isInitialized -> {
+                if (nextRoom.contains(currentPosition)) {
+                    currRoom = nextRoom
+                }
+                /*
+                 * We should have reached nextRoom, but we are in a different room (newRoom),
+                 * in such case we return back to the previous room (stored in currRoom), so
+                 * as to correctly follow our route.
+                 * In the future, something more sophisticated could be done (e.g. recomputing
+                 * the route).
+                 */
+                else {
+                    val prevRoom = currRoom
+                    currRoom = newRoom
+                    val doorsLeadingBack = environment.graph().outgoingEdgesOf(currRoom)
+                        .filter { environment.graph().getEdgeTarget(it) == prevRoom }
+                    moveToClosestDoor(doors = doorsLeadingBack)
+                    return
+                }
+            }
+            else -> currRoom = newRoom
```

Comment:
`when` is functional and should not get used imperatively. Any reason not to use `if` here?